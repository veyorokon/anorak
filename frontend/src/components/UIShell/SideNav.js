/**
 * Copyright IBM Corp. 2016, 2018
 *
 * This source code is licensed under the Apache-2.0 license found in the
 * LICENSE file in the root directory of this source tree.
 */

import React, { useState, useRef } from 'react';
import { settings } from 'carbon-components';
import cx from 'classnames';
import PropTypes from 'prop-types';
import { AriaLabelPropType } from '../../prop-types/AriaPropTypes';
import SideNavFooter from './SideNavFooter';

const { prefix } = settings;

const SideNav = React.forwardRef(function SideNav(props, ref) {
  const {
    expanded: expandedProp,
    defaultExpanded,
    'aria-label': ariaLabel,
    'aria-labelledby': ariaLabelledBy,
    children,
    onToggle,
    className: customClassName,
    translateById: t,
  } = props;
  const { current: controlled } = useRef(expandedProp !== undefined);
  const [expandedState, setExpandedState] = useState(defaultExpanded);
  const expanded = controlled ? expandedProp : expandedState;

  const handleToggle = (event, value = !expanded) => {
    if (!controlled) {
      setExpandedState(value);
    }
    if (onToggle) {
      onToggle(event, value);
    }
  };

  const accessibilityLabel = {
    'aria-label': ariaLabel,
    'aria-labelledby': ariaLabelledBy,
  };

  const assistiveText = expanded
    ? t('carbon.sidenav.state.open')
    : t('carbon.sidenav.state.closed');

  const className = cx({
    [`${prefix}--side-nav`]: true,
    [`${prefix}--side-nav--expanded`]: expanded,
    [customClassName]: !!customClassName,
  });

  return (
    <nav
      ref={ref}
      className={`${prefix}--side-nav__navigation ${className}`}
      {...accessibilityLabel}
      onFocus={event => handleToggle(event, true)}
      onBlur={event => handleToggle(event, false)}>
      {children}
      <SideNavFooter
        assistiveText={assistiveText}
        expanded={expanded}
        onToggle={handleToggle}
      />
    </nav>
  );
});

SideNav.defaultProps = {
  translateById: id => {
    const translations = {
      'carbon.sidenav.state.open': 'Close',
      'carbon.sidenav.state.closed': 'Open',
    };
    return translations[id];
  },
  defaultExpanded: false,
};

SideNav.propTypes = {
  /**
   * If `true`, the SideNav will be expanded, otherwise it will be collapsed.
   * Using this prop causes SideNav to become a controled component.
   */
  expanded: PropTypes.bool,

  /**
   * If `true`, the SideNav will be open on initial render.
   */
  defaultExpanded: PropTypes.bool,

  /**
   * An optional listener that is called when an event that would cause
   * toggling the SideNav occurs.
   *
   * @param {object} event
   * @param {boolean} value
   */
  onToggle: PropTypes.func,

  /**
   * Required props for accessibility label on the underlying menu
   */
  ...AriaLabelPropType,

  /**
   * Optionally provide a custom class to apply to the underlying <li> node
   */
  className: PropTypes.string,

  /**
   * Provide a custom function for translating all message ids within this
   * component. This function will take in two arguments: the mesasge Id and the
   * state of the component. From this, you should return a string representing
   * the label you want displayed or read by screen readers.
   */
  translateById: PropTypes.func,
};

export default SideNav;
