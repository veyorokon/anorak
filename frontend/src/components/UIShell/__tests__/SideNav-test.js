/**
 * Copyright IBM Corp. 2016, 2018
 *
 * This source code is licensed under the Apache-2.0 license found in the
 * LICENSE file in the root directory of this source tree.
 */

import React from 'react';
import { mount } from 'enzyme';
import SideNav from '../SideNav';

describe('SideNav', () => {
  let mockProps;

  beforeEach(() => {
    mockProps = {
      'aria-label': 'Navigation',
      children: <h2>Navigation</h2>,
    };
  });

  it('should render', () => {
    const wrapper = mount(<SideNav {...mockProps} />);
    expect(wrapper).toMatchSnapshot();
  });
});
