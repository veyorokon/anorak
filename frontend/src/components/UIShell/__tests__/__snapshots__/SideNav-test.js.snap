// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SideNav should render 1`] = `
<ForwardRef(SideNav)
  aria-label="Navigation"
  defaultExpanded={false}
  translateById={[Function]}
>
  <nav
    aria-label="Navigation"
    className="bx--side-nav__navigation bx--side-nav"
    onBlur={[Function]}
    onFocus={[Function]}
  >
    <h2>
      Navigation
    </h2>
    <SideNavFooter
      assistiveText="Open"
      expanded={false}
      onToggle={[Function]}
    >
      <footer
        className="bx--side-nav__footer"
      >
        <button
          className="bx--side-nav__toggle"
          onClick={[Function]}
          title="Open"
          type="button"
        >
          <div
            className="bx--side-nav__icon"
          >
            <ChevronRight20
              height={20}
              preserveAspectRatio="xMidYMid meet"
              viewBox="0 0 32 32"
              width={20}
              xmlns="http://www.w3.org/2000/svg"
            >
              <svg
                aria-hidden={true}
                focusable="false"
                height={20}
                preserveAspectRatio="xMidYMid meet"
                style={
                  Object {
                    "willChange": "transform",
                  }
                }
                viewBox="0 0 32 32"
                width={20}
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M22 16L12 26l-1.4-1.4 8.6-8.6-8.6-8.6L12 6z"
                />
              </svg>
            </ChevronRight20>
          </div>
          <span
            className="bx--assistive-text"
          >
            Open
          </span>
        </button>
      </footer>
    </SideNavFooter>
  </nav>
</ForwardRef(SideNav)>
`;
